Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -774,20 +774,124 @@ namespace OpenInfraPlatform {
+						}
+						else
+						{
+							// general case: arbitrary number of vertices, possible inner bound (= hole)
+							//  -> elaborate triangulation with respect to arbitrary number of vertices and holes is necessary;
+
+							addArbitraryFaceToPolyhedronData(face, faceBoundLoops, polygon, polygonIndices);
+						}
+					}
+				}
+
+				void convertIfcFaceSurface(
+					const EXPRESSReference<typename IfcEntityTypesT::IfcFaceSurface>& faceSurface,
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops,
+					const carve::math::Matrix& pos,
+					std::shared_ptr<carve::input::PolyhedronData>& polygon, //Carve polygon of the converted face
+					std::map<std::string, uint32_t> polygonIndices // Contains polygon indices of vertices (x,y,z converted to string)
+				)  const noexcept(false)
+				{
+					if (faceSurface.expired()) {
+						throw oip::ReferenceExpiredException(faceSurface);
+					}
+
+					if (faceSurface.isOfType<typename IfcEntityTypesT::IfcAdvancedFace>())
+					{
+						convertIfcAdvancedFace(faceSurface.as<typename IfcEntityTypesT::IfcAdvancedFace>(), faceBoundLoops, pos, polygon, polygonIndices);
+					}
+					else
+					{
+						throw oip::UnhandledException(faceSurface);
+					}
+				}
+
+				void convertIfcAdvancedFace(
+					const EXPRESSReference<typename IfcEntityTypesT::IfcAdvancedFace>& advancedFace,
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops,
+					const carve::math::Matrix& pos,
+					std::shared_ptr<carve::input::PolyhedronData>& polygon, //Carve polygon of the converted face
+					std::map<std::string, uint32_t> polygonIndices // Contains polygon indices of vertices (x,y,z converted to string)
+				)  const noexcept(false)
+				{
+					if (advancedFace.expired()) {
+						throw oip::ReferenceExpiredException(advancedFace);
+					}
+
+					// get reference to IfcFaceBound's (attribute 1)
+					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>>& ifcFaceBounds = advancedFace->Bounds;
+					// if one face bound is not of the allowed type: throw; else: continue and use already converted faceBoundLoops
+					if (!std::all_of(ifcFaceBounds.begin(), ifcFaceBounds.end(), [](const auto& ifcFaceBound) {return
+						ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcEdgeLoop>() || ifcFaceBound->Bound.isOfType<typename IfcEntityTypesT::IfcVertexLoop>(); }))
+					{
+						throw oip::InconsistentModellingException(advancedFace, "IfcAdvancedFace has a loop type as face boundary which is not allowed.");
+					}
+
+					// get attribute 3: indicates whether the sense of the surface normal agrees with the sense of the topological normal (face bound)
+					// ToDo: take sameSense into account by the addFace-construction
+					const bool sameSense = advancedFace->SameSense;
+
+					// get attribute 2
+					const EXPRESSReference<typename IfcEntityTypesT::IfcSurface>& surface = advancedFace->FaceSurface;

```

Comment:
Is there no `convertIfcSurface` available? If not, please introduce one.

Reply:
This section is moved into the new function computeIfcFaceSurface (commit b9d8d5d).
Inside there, the surface is needed as EXPRESSReference, because its surface-type (IfcPlane or everything else) is part of the decision about the computation approach. The big if-else construct was simplified in commit 6c3e76d.